### DISCUSSION TITLE
atomFamily with atomWithDefault and useResetAtom

### DISCUSSION THREAD

**Discussion by janrembold** [pmndrs/jotai#1400] [General] [answered]
I'm trying to implement the following features and (in theory) this should be quite simple, but I don't get it to work.

- Using `atomFamily<string, Promise<string>>`
- Setting the default value with `atomWithDefault`
- Being able to reset the atom with `useResetAtom` what triggers the default function from `atomWithDefault`

Codesandbox: https://codesandbox.io/s/jotai-atomfamily-with-atomwithdefault-txsr6e?file=/src/App.tsx 

``` typescript
// this is the (not working) idea 
const getThePromise = (id: string): Promise<string> =>
  Promise.resolve(`Baz #${id} - ${Date.now()}`);

export const familyAtomWithDefault = atomFamily<string, Promise<string>>((id) =>
  atomWithDefault(() => getThePromise(id))
);
```

Can you kick me in the right direction please?! I think this would be some basic task for a family with default values, but haven't found any working examples out there...

Thanks in advance!!!

**Comment by dai-shi** [comment, marked as answer]
Can you check this?
https://codesandbox.io/s/jotai-atomfamily-with-atomwithdefault-forked-2z5w9y?file=/src/App.tsx

**Reply by janrembold** [reply]
At least this resolves the TypeScript errors and `useAtom` is working now. 
But the reset button is still dead :(

**Reply by dai-shi** [reply]
> the reset button is still dead

Can you reproduce the issue without atomFamily?

**Reply by janrembold** [reply]
The `atomWithDefault` reset works like expected: https://codesandbox.io/s/jotai-atomfamily-with-atomwithdefault-forked-7miy62?file=/src/App.tsx

**Reply by dai-shi** [reply]
I just added `atomFamily` to it: https://codesandbox.io/s/jotai-atomfamily-with-atomwithdefault-forked-10ifyb?file=/src/App.tsx

**Reply by janrembold** [reply]
Oh, I think I see the problem. The equal check might be the problem here. The reset only works (of course) when there was a change in the value. So resetting the default value does not trigger a re-render. 

I'm not familiar with how those resets are working behind the scenes but it looks to me that the equality check just doesn't see a change as the default value is a `Promise` and the new value is also a `Promise`. 

In my real world application this wouldn't be a simple `Promise.resolve('string123')` but a fetch request Promise. Don't know if this makes a difference for the equal check but as the value hasn't changed it should still remain as unchanged and therefore not reset itself

**Reply by dai-shi** [reply]
Right, resetting doesn't mean refetching. Atoms are only re-evaluated when its dependencies change. Because it has no dependencies, it will never change. It's declarative.

I think you want imperative approach, something like this.
```js
export const familyAtom = atomFamily((id: string) =>
  atom(getThePromise(id))
);

  const [foo2, setFoo2] = useAtom(familyAtomWithDefault("foo2"));
  const reset2 = () => setFoo2(getThePromise("foo2"));
```

See also: https://jotai.org/docs/guides/async#async-sometimes

If you want declarative approach and still want to refetch, there's a hack solution: https://jotai.org/docs/advanced-recipes/atom-creators#atom-with-refresh
